### PR TITLE
test: add PAM functional testing for Azure Linux

### PR DESCRIPTION
**Pull request by corvus-callidus**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:
During the course of investigation for the CIS benchmark, it was discovered that while the PAM configuration for Azure Linux passed the existing tests, it didn't actually work as expected. Line order is critical within a PAM config and impacts function. This change adds tests to ensure the PAM configuration for Mariner is functional and in alignment with our CIS benchmark and Azure security requirements. 

The fixes for the PAM configuration were addressed in #3248

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+      username (str): The name of the user to set the password for;
+      password (str): The password to set for the user
+    """
+    print("\n=== Setting password for " + username + " to " + password)
+    child = BashSpawn(f"sudo passwd {username}", encoding='utf-8',
+                      logfile=sys.stdout)
+    try:
+        # child.expect_optional_sudo_auth()
+        child.expect_password_prompt(password)
+        if child.expect(["Retype new password:"]) == 0:
```
Comment (tobiasb-ms) [inline vhdbuilder/packer/test/pam/test_pam.py:109]: You can get rid of the `if` and `else` here -- `expect` will raise an exception if it doesn't find a match.

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+import pexpect
+import pytest
+
+
+class BashSpawn(pexpect.spawn):
+    """
+    Extension of pexpect.spawn to handle bash, login and password prompts
+    """
+    def __init(self, command, args=[]):
+
```
Comment (tobiasb-ms) [inline vhdbuilder/packer/test/pam/test_pam.py:16]: nit: unless there's some idiomatic python thing here, remove this blank line

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+        print("\n-----logged in as " + user.name)
+
+        # attempt to change password
+        child.sendline("passwd")
+        print("\n-----sending current password")
+        child.expect_password_prompt(user.pw)
+
+        done = False
+        while not done:
+            i = child.expect(["[Nn]ew.*password:",
```
Comment (tobiasb-ms) [inline vhdbuilder/packer/test/pam/test_pam.py:368]: I think this will raise an exception if the password is ever changed -- is that intentional?

**Review by tobiasb-ms — APPROVED** [review]
Great work on this. I love the move toward a more fully-featured test framework!

File: vhdbuilder/packer/test/pam/test_pam.py (hunk 3 above)
Comment (corvus-callidus) [reply vhdbuilder/packer/test/pam/test_pam.py:368]: It's only ever used in an "expected to fail" context, so yes.

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+
+    Input Arguments:
+      username (str): The name of the user to set the password for;
+      password (str): The password to set for the user
+    """
+    print("\n=== Setting password for " + username + " to " + password)
+    child = BashSpawn(f"sudo passwd {username}", encoding='utf-8',
+                      logfile=sys.stdout)
+    try:
+        # child.expect_optional_sudo_auth()
```
Comment (alexeldeib) [inline vhdbuilder/packer/test/pam/test_pam.py:106]: leftover?

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+        Raises:
+          Exception: If the user is not deleted successfully or the faillock
+          file is not removed successfully
+        """
+
+        print(f"\n=== Deleting user {self.name}")
+        child = BashSpawn(f"sudo userdel -f {self.name}", encoding='utf-8')
+        try:
+            # child.expect_optional_sudo_auth()
```
Comment (alexeldeib) [inline vhdbuilder/packer/test/pam/test_pam.py:180]: leftover?

File: vhdbuilder/packer/test/pam/test_pam.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+            print("\n-----Failed to delete user: " + str(e))
+            pytest.fail("Failed to delete user")
+        finally:
+            child.close()
+
+        print(f"\n-----Removing faillock file for {self.name}")
+        child = BashSpawn(f"sudo rm -f /var/run/faillock/{self.name}",
+                          encoding='utf-8')
+        try:
+            # child.expect_optional_sudo_auth()
```
Comment (alexeldeib) [inline vhdbuilder/packer/test/pam/test_pam.py:198]: 👀

File: vhdbuilder/packer/test/pam/test_pam.py (hunk 4 above)
Comment (corvus-callidus) [reply vhdbuilder/packer/test/pam/test_pam.py]: Oops. Fixed now, as are all the rest.

**Review by alexeldeib — APPROVED** [review]

**Pull request merged**